src/pybind/mgr/mgr_module.py

```diff
@@ -1731,6 +1731,62 @@ def tell_command(self, daemon_type: str, daemon_id: str, cmd_dict: dict, inbuf:
         ))
 
         return r
+    
+    MDS_STATE_ORD = {
+        "down:dne":              0, # CEPH_MDS_STATE_DNE,
+        "down:stopped":         -1, # CEPH_MDS_STATE_STOPPED,
+        "down:damaged":         15, # CEPH_MDS_STATE_DAMAGED,
+        "up:boot":              -4, # CEPH_MDS_STATE_BOOT,
+        "up:standby":           -5, # CEPH_MDS_STATE_STANDBY,
+        "up:standby-replay":    -8, # CEPH_MDS_STATE_STANDBY_REPLAY,
+        "up:oneshot-replay":    -9, # CEPH_MDS_STATE_REPLAYONCE,
+        "up:creating":          -6, # CEPH_MDS_STATE_CREATING,
+        "up:starting":          -7, # CEPH_MDS_STATE_STARTING,
+        "up:replay":             8, # CEPH_MDS_STATE_REPLAY,
+        "up:resolve":            9, # CEPH_MDS_STATE_RESOLVE,
+        "up:reconnect":         10, # CEPH_MDS_STATE_RECONNECT,
+        "up:rejoin":            11, # CEPH_MDS_STATE_REJOIN,
+        "up:clientreplay":      12, # CEPH_MDS_STATE_CLIENTREPLAY,
+        "up:active":            13, # CEPH_MDS_STATE_ACTIVE,
+        "up:stopping":          14, # CEPH_MDS_STATE_STOPPING,
+    }
+    MDS_STATE_ACTIVE_ORD = MDS_STATE_ORD["up:active"]
+
+    def get_quiesce_leader_info(self, fscid: str) -> Optional[dict]:
+        leader_info: Optional[dict] = None
+
+        for fs in self.get("fs_map")['filesystems']:
+            if fscid != fs["id"]:
+                continue
+            
+            # quiesce leader is the lowest rank
+            # with the highest state
+            mdsmap = fs["mdsmap"]
+            for info in mdsmap['info'].values():
+                if info['rank'] == -1:
+                    continue
+                if leader_info is None:
+                    leader_info = info
+                else:
+                    if info['rank'] < leader_info['rank']:
+                        leader_info = info
+                    elif info['rank'] == leader_info['rank']:
+                        state_ord = self.MDS_STATE_ORD.get(info['state'])
+                        # if there are more than one daemons with the same rank
+                        # only one of them can be active
+                        if state_ord == self.MDS_STATE_ACTIVE_ORD:
+                            leader_info = info
+            break
+
+        return leader_info
```

Comment:
I'll respond to some of the bullet points here, but I will also try and cover the relevant bits in the code comments as you request. 

- What happens if we send a command to a rank that is no longer the leader? Should the command be retried with a new MDSMap?
  - Currently, the user will receive an ENOTTY error and will have to retry. We could add a local retry after synchronizing the MDSMap, but I suggest that we leave this for a later enhancement 
- Can only the leader respond to all quiesce db commands?
  - Yes, only the leader is authorized to respond to quiesce db commands, even for queries.
- Why not ask the MDS who the leader is and then send a follow-up db command to the leader?
  - This would be against the design, which is to deduce the leader from the MDSMap. This is a cornerstone of the quiesce db clustering approach, where we offload this decision to the MDSMap. Ideally, the MDSMonitor would perform this calculation locally and expose the result as a dedicated field in the MDSMap.
- If a degraded cluster [cannot quiesce](https://github.com/ceph/ceph/pull/54485/files#diff-d4d03b797940276595f7cad04a78b63c56440ee19f3c5c5336a5994a5a2217fcR490-R494) then why not have the client always wait until mds.0 is up:active (and therefore the leader) before sending a command
  - whether a degraded cluster can or can't quiesce is not for the client to decide, but only for the MDS. I expect us to lift this limitation in the future, or at least put more effort into quiescing roots that aren't affected by the missing ranks.

> I do want to add that I'm overall impressed with how compact and simple the volumes plugin glue is. That is commendable.

Thank you, I was happy about it as well! It helped a lot when you suggested that we just send the command as is to the admin socket, otherwise there would be much more work to expose the quiesce db model via a c API from the libcephfs.

Reply:
> I'll respond to some of the bullet points here, but I will also try and cover the relevant bits in the code comments as you request.
> 
>     * What happens if we send a command to a rank that is no longer the leader? Should the command be retried with a new MDSMap?
>       
>       * Currently, the user will receive an ENOTTY error and will have to retry. We could add a local retry after synchronizing the MDSMap, but I suggest that we leave this for a later enhancement

I would like to see a simple (iteration limited) retry. Resending commands to the mgr is not cheap.

>     * Can only the leader respond to all quiesce db commands?
>       
>       * Yes, only the leader is authorized to respond to quiesce db commands, even for queries.

Please note that somewhere in the tell handler and in the py code.

>     * Why not ask the MDS who the leader is and then send a follow-up db command to the leader?
>       
>       * This would be against the design, which is to deduce the leader from the MDSMap. This is a cornerstone of the quiesce db clustering approach, where we offload this decision to the MDSMap. Ideally, the MDSMonitor would perform this calculation locally and expose the result as a dedicated field in the MDSMap.

Updating the MDSMap for something that can be deduced is undesirable. I'm not thrilled with hard-coding the leader algorithm in the volumes plugin, especially when it means dragging a bunch of state enums into the python code . Ultimately leader selection is the MDS's job. It just is currently done via a calculation on the MDSMap which to me is an implementation detail the client should be insulated from.
 
>     * If a degraded cluster [cannot quiesce](https://github.com/ceph/ceph/pull/54485/files#diff-d4d03b797940276595f7cad04a78b63c56440ee19f3c5c5336a5994a5a2217fcR490-R494) then why not have the client always wait until mds.0 is up:active (and therefore the leader) before sending a command
>       
>       * whether a degraded cluster can or can't quiesce is not for the client to decide, but only for the MDS.

... but the client should be able to compute the leader independently? We're going in circles here but I feel you have designed a leader-replica system for a distributed system that does not exist and is very unlikely to exist with the current architecture. I am okay with replicating the database back to the leader to avoid "losing" it unnecessarily; but, this code is creating complexity in the client/MDS to maintain a leader of a degraded file system when all outstanding quiesce sets will ultimately be cancelled and new quiesce sets cannot be made until the file system is again healthy (where rank 0 is again active). For example, this code code could simply check whether rank 0 is `up:active` and raise an exception if not. Then CSI could retry the whole quiesce/snapshot operation at a future time.

Reply:
The original design that this emulates is that MDSMap assigns a quiesce leader to whom communications should be directed. That's it. I don't see any conflict here with the current design. This approach is the same as how rank X is messaged today by looking its addresses up in an MDS map. 

> Updating the MDSMap for something that can be deduced is undesirable

The current implementation avoids this and we see how it creates code duplication. Even when we eliminate the duplication by using a helper function in a common header, it is not upgrade-safe if the logic is to be updated between versions.

> but, this code is creating complexity in the client/MDS to maintain a leader of a degraded file system 

The fact that some other process today is limited to the availability of rank 0 is not an argument that this subsystem should artificially impose the same limitation. Figuring out who to message is a trivial task that client will have to perform one way or another.

Reply:
Addressed by implementing the leader selection logic on the MDSMonitor side. See latest code.